File: Python/0234_palindrome_linked_list.py
# 2020-08-05 14:47:46

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        len_ = 0
        if head is None or head.next is None:
            return True
        
        p = head
        while p is not None:
            len_ += 1
            p = p.next

        prev = None
        cur = head
        for i in range(len_ // 2):
            next_ = cur.next
            cur.next = prev
            prev = cur
            cur = next_

        if (len_ & 1) == 1:
            cur = cur.next

        while cur is not None and prev is not None:
            if cur.val != prev.val:
                return False
            cur = cur.next
            prev = prev.next

        return True

File: TypeScript/0234-palindrome-linked-list.ts
// 2020-08-05 17:33:18

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function isPalindrome(head: ListNode | null): boolean {
  if (head == null || head.next == null) return true;
  let len = 0;
  let p: ListNode | null = head;
  for (; p != null; p = p.next) { len++; }

  let prev: ListNode | null = null;
  let cur: ListNode | null = head;
  for (let i = 0; i < Math.floor(len / 2); i++) { 
    let next: ListNode | null = cur!.next;
    cur!.next = prev;
    prev = cur;
    cur = next;
  }

  if (len % 2 == 1) cur = cur!.next;

  while (prev != null && cur != null) {
    if (prev.val != cur.val) return false;
    prev = prev.next;
    cur = cur.next;
  }
  return true;

};